GVFS/GVFS.Common/Maintenance/GitMaintenanceStep.cs

```diff
@@ -114,27 +114,37 @@ public void Stop()
         /// </summary>
         protected abstract void PerformMaintenance();
 
-        protected GitProcess.Result RunGitCommand(Func<GitProcess, GitProcess.Result> work)
+        protected GitProcess.Result RunGitCommand(Func<GitProcess, GitProcess.Result> work, string gitCommand)
         {
-            using (ITracer activity = this.Context.Tracer.StartActivity("RunGitCommand", EventLevel.Informational, Keywords.Telemetry, metadata: this.CreateEventMetadata()))
+            EventMetadata metadata = this.CreateEventMetadata();
+            metadata.Add("gitCommand", gitCommand);
+
+            using (ITracer activity = this.Context.Tracer.StartActivity("RunGitCommand", EventLevel.Informational, Keywords.Telemetry, metadata))
             {
                 if (this.Stopping)
                 {
-                    string message = this.Area + "Not launching Git process because the mount is stopping";
+                    string message = $"{this.Area}: Not launching Git process {gitCommand} because the mount is stopping";
                     this.Context.Tracer.RelatedWarning(
                         metadata: null,
                         message: message,
                         keywords: Keywords.Telemetry);
                     return new GitProcess.Result(message, string.Empty, GitProcess.Result.ExitDueToShutDownCode);
                 }
 
-                GitProcess.Result result = work.Invoke(this.GitProcess);
+                GitProcess.Result result = work.Invoke(this.MaintenanceGitProcess);
 
                 if (!this.Stopping && result?.ExitCodeIsFailure == true)
                 {
+                    string errorMessage = result?.Errors == null ? string.Empty : result.Errors;
+                    if (errorMessage.Length > 1000)
+                    {
+                        // For large error messages, we show the first and last 500 chars
+                        errorMessage = $"beginning: {errorMessage.Substring(0, 500)} ending: {errorMessage.Substring(errorMessage.Length - 500)}";
+                    }

```

Comment:
Should this be moved to the tracer or do we only expect large error messages from git?

Reply:
We only expect the large error messages from git.

If we think we could have an epidemic of large messages we could limit message size coming from the tracer directly the same way. I would punt that to the future though, since that affects all messages and will require more thought.